### PR TITLE
Inflight nemo model export support

### PR DESCRIPTION
**Pull request by JimmyZhang12**
# What does this PR do ?
Adds inflight NeMo to TRTLLM v10 model conversion and engine refitting using device weights

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/export/tensorrt_llm.py [hunk 1]
```diff
@@ -317,6 +325,115 @@ def export(
         if load_model:
             self._load()
 
+    def build(
```
Comment (oyilmaz-nvidia) [inline nemo/export/tensorrt_llm.py:328]: How is this different then the export? It looks similar to export. If yes, can we merge this function with the export function? We might add a couple of parameters to control anything specific with refit.

File: nemo/export/trt_llm/converter/model_converter.py [hunk 2]
```diff
@@ -83,30 +89,40 @@ def model_to_trtllm_ckpt(
...
-    )
-
-    world_size = tensor_parallel_size * pipeline_parallel_size
-
-    has_lm_head = "lm_head.weight" in weights_dict
-    if has_lm_head:
-        lm_head_weight = weights_dict["lm_head.weight"]
+    # If the model has been sharded with model parallelism, convert the model in a gpu-distributed manner
+    if use_distributed_convert:
+        weights_dict = dist_model_to_trt_llm_ckpt(
```
Comment (oyilmaz-nvidia) [inline nemo/export/trt_llm/converter/model_converter.py:94]: Is this for the old nemo ckpt type? Old one wasn't called the dist ckpt btw. And we no longer support the old ckpt format.

**Review by oyilmaz-nvidia — COMMENTED** [review]
Why do you need the support for the old nemo ckpt format?

File: nemo/export/trt_llm/converter/model_converter.py (hunk 2 above)
Comment (jiemingz) [reply nemo/export/trt_llm/converter/model_converter.py:94]: This converts an instantiated nemo model that may be sharded with model parallelism on device, and not a checkpoint

Comment (oyilmaz-nvidia) [reply nemo/export/trt_llm/converter/model_converter.py:94]: You mean this is for the models that are already in the GPU memory?

File: nemo/export/tensorrt_llm.py (hunk 1 above)
Comment (jiemingz) [reply nemo/export/tensorrt_llm.py:328]: For instance: https://github.com/NVIDIA/NeMo/blob/a973ef529f98c90be5825fac4bcaca869f25e294/nemo/export/tensorrt_llm.py#L243
In export a single rank creates the engines in a loop, but in build all gpus create engines simultaneously (because the distributed environment has already been created)

File: nemo/export/trt_llm/converter/model_converter.py (hunk 2 above)
Comment (jiemingz) [reply nemo/export/trt_llm/converter/model_converter.py:94]: Yeah exactly.

Comment (jiemingz) [reply nemo/export/trt_llm/converter/model_converter.py:94]: It's also why we add new arguments like the vocab size, since we need to unpad the embedding layers, which NeMo pads upon creating the model from the checkpoint.

Comment (oyilmaz-nvidia) [reply nemo/export/trt_llm/converter/model_converter.py:94]: Can you please add this refit option to this script https://github.com/NVIDIA/NeMo/blob/main/scripts/deploy/nlp/deploy_triton.py and this test https://github.com/NVIDIA/NeMo/blob/main/tests/export/nemo_export.py ? Let me take a look how it works.

File: nemo/export/trt_llm/converter/model_to_trt_llm_ckpt.py [hunk 3]
```diff
@@ -238,6 +239,246 @@ def handle_model_level_weights(model, tp_idx: int, pp_idx: int):
     return weights_dict
 
 
+def _get_layer_index(split_key):
```
Comment (oyilmaz-nvidia) [inline nemo/export/trt_llm/converter/model_to_trt_llm_ckpt.py:242]: There are too many duplicate code here. Can we reuse some of the existing functions and add parameters and code as needed?

File: nemo/export/trt_llm/converter/utils.py [hunk 4]
```diff
@@ -414,3 +414,124 @@ def split(v, tp_size, idx, dim=0):
         return np.ascontiguousarray(np.split(v, tp_size)[idx])
     else:
         return np.ascontiguousarray(np.split(v, tp_size, axis=dim)[idx])
+
+
+# Similar to split_save_weight but done on GPU for performance
+@torch.no_grad()
+def split_save_weight_gpu(key, val, config):
```
Comment (oyilmaz-nvidia) [inline nemo/export/trt_llm/converter/utils.py:421]: There are also too many duplicate code here. Can we reuse some of the existing functions and add parameters and code as needed?

**Review by oyilmaz-nvidia — CHANGES_REQUESTED** [review]

File: nemo/export/trt_llm/converter/model_to_trt_llm_ckpt.py (hunk 3 above)
Comment (JimmyZhang12) [reply nemo/export/trt_llm/converter/model_to_trt_llm_ckpt.py:242]: Can you please carefully review this code? To my knowledge this function is not duplicated anywhere

Comment (oyilmaz-nvidia) [reply nemo/export/trt_llm/converter/model_to_trt_llm_ckpt.py:242]: 1. Didn't mean specifically _get_layer_index function.
2. Also, I don't mean the exact copy of the code by duplicate but rather code with similar logic.

Comment (JimmyZhang12) [reply nemo/export/trt_llm/converter/model_to_trt_llm_ckpt.py:242]: The convention export currently uses ([see here](https://github.com/NVIDIA/NeMo/blob/7fee8e7a0f576317e4113cd58282fa833358c574/nemo/export/trt_llm/converter/utils.py#L179)) is to hardcode index 1. For Nemo models the layer number is not guaranteed (for instance if the module has been wrapped in autocast) so we have unique logic to determine the layer index.

File: nemo/export/trt_llm/converter/model_to_trt_llm_ckpt.py [hunk 5]
```diff
@@ -238,6 +239,246 @@ def handle_model_level_weights(model, tp_idx: int, pp_idx: int):
...
+
+
+def get_layer_num(param_name):
+    split_key = param_name.split(".")
+    layer_index = int(_get_layer_index(split_key))
+    return int(split_key[layer_index])
+
+
+@torch.no_grad()
+def dist_model_to_trt_llm_ckpt(
```
Comment (oyilmaz-nvidia) [inline nemo/export/trt_llm/converter/model_to_trt_llm_ckpt.py:263]: Can dist_model_to_trt_llm_ckpt function and convert_model_to_trt_llm_ckpt be merged into one function for instance?

**Review by oyilmaz-nvidia — APPROVED** [review]
Approving with the condition to work on this again after the release to provide one export path for all use cases.

**Review by jiemingz — APPROVED** [review]
